File: solutions/ruby/knapsack/2/knapsack.rb
require 'matrix'

class Item
    attr_reader :weight, :value
    def initialize(weight, value)
        @weight = weight
        @value = value
    end
end

class Knapsack
    def initialize(max_weight)
        @max_weight = max_weight
    end

    def max_value(items)
        dp = Matrix.build(items.length + 1, @max_weight + 1) { 0 }

        (1..items.length).each do |i|
            (1..@max_weight).each do |w|
                item = items[i - 1]
                if item.weight > w
                    dp[i, w] = dp[i - 1, w]
                else
                    dp[i, w] = [dp[i - 1, w], dp[i - 1, w - item.weight] + item.value].max
                end
            end
        end

        dp[items.length, @max_weight]
    end
end

File: solutions/elixir/knapsack/2/lib/knapsack.ex
defmodule Knapsack do
  @doc """
  Return the maximum value that a knapsack can carry.
  """
  @spec maximum_value(items :: [%{value: integer, weight: integer}], maximum_weight :: integer) ::
          integer
  def maximum_value([], _), do: 0

  def maximum_value([%{value: value, weight: weight} | tail], maximum_weight) do
    cond do
      weight > maximum_weight ->
        maximum_value(tail, maximum_weight)

      true ->
        max(
          value + maximum_value(tail, maximum_weight - weight),
          maximum_value(tail, maximum_weight)
        )
    end
  end
end
